### PR TITLE
Fix test flakiness with OnTimeout

### PR DESCRIPTION
**Pull request by BrennanConroy**
Fixes part of https://github.com/dotnet/aspnetcore/issues/21520
Fixes https://github.com/dotnet/aspnetcore-internal/issues/2197

Before, if you froze the thread right after calling `_flusher.FlushAsync` and before `_timeoutControl.BytesWrittenToBuffer` on the last write before the timeout check in the tests then you could 100% of the time repro the test issue. I can't recall the exact reasons because it's been awhile since I read through the code, but after moving the `BytesWrittenToBuffer` before flushing, the test no longer fails.

This should be fine because we weren't awaiting the flush before calling the timeout controller, so we haven't really changed any behavior here.

### CODE REVIEW

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs [hunk 1]
```diff
@@ -439,24 +439,29 @@ private async ValueTask<FlushResult> WriteDataAsync(int streamId, StreamOutputFl
...
                     }
 
                     firstWrite = false;
 
-                    if (_minResponseDataRate != null)
-                    {
-                        _timeoutControl.BytesWrittenToBuffer(_minResponseDataRate, _unflushedBytes);
-                    }
-
                     _unflushedBytes = 0;
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:465]: This should only be reset to zero if we've called BytesWrittenToBuffer or _minResponseDataRate is null.

File: src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs [hunk 2]
```diff
@@ -439,24 +439,29 @@ private async ValueTask<FlushResult> WriteDataAsync(int streamId, StreamOutputFl
                             dataLength = 0;
                         }
 
+                        if (_minResponseDataRate != null)
+                        {
+                            _timeoutControl.BytesWrittenToBuffer(_minResponseDataRate, _unflushedBytes);
```
Comment (halter73) [inline src/Servers/Kestrel/Core/src/Internal/Http2/Http2FrameWriter.cs:444]: We should add a comment indicating that we call into ITimeoutControl.BytesWrittenToBuffer before flushing the output pipe for testing reasons.

**Comment by BrennanConroy** [comment]
Bump

File: src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs [hunk 3]
```diff
@@ -582,7 +583,7 @@ public async Task DATA_Sent_TooSlowlyDueToOutputFlowControlOnMultipleStreams_Abo
             await SendDataAsync(3, _maxData, endStream: true);
 
             await ExpectAsync(Http2FrameType.HEADERS,
-                withLength: 32,
+                withLength: 2,
```
Comment (jkotalik) [inline src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs:586]: Why did the length change?

Comment (BrennanConroy) [reply src/Servers/Kestrel/test/InMemory.FunctionalTests/Http2/Http2TimeoutTests.cs:586]: This test was likely written before we added some HTTP2 feature that allowed this to be smaller. Other tests in this file that expect 2 HEADERS frames also expect length 2 for the second one.

**Review by halter73 — APPROVED** [review]

**Review by jkotalik — APPROVED** [review]

**Pull request merged**